File: 2020/72.py
"""
see https://leetcode.com/problems/edit-distance/
"""
class Solution(object):
    def minDistance(self, w1, w2):
        """
        :type word1: str
        :type word2: str
        :rtype: int
        """
        # Consider w1(abc) -> w2(defg)
        # At most we can use len(w2) steps to to complete the conversion

        # Use DP
        # say len(w1) = l1, len(w2) = l2
        # create an 2D-array
        # which arr[i][j] represents the required steps for w1[:i+1] -> w2[:j+1]
        # where 0 < i < len(w1), 0 < j < len(w2)
        # initial state: dp[i][j] = j
        # then we evaluate the conversion for l1 substrings -> l2 substrings
        # keep update arr

        #   "" h o r s e      
        # "" 0 1 2 3 4 5  => Deleting these many chars from Horse to make ""
        # r  1 1 2 2 3 4 
        # o  2 2 1 2 3 4 
        # s  3 3 2 2 2 3 
        #    ||
        #    insert these many chars to make a empty string to ros

        dp = [[0 for j in range(len(w2)+1)] for i in range(len(w1)+1)]

        for i in range(len(w1)+1):
            for j in range(len(w2)+1):
                if i == 0:
                    dp[i][j] = j
                elif j == 0:
                    dp[i][j] = i
                elif w1[i-1] == w2[j-1]:
                    dp[i][j] = dp[i-1][j-1]
                else:
                    dp[i][j] = min(dp[i-1][j-1], dp[i-1][j], dp[i][j-1])+1

        return dp[-1][-1]

print(Solution().minDistance("og", "grs")==3)
print(Solution().minDistance("intention", "execution")==5)
print(Solution().minDistance("sea", "eat") == 2)

File: current_session/c++/72.cpp
#include <iostream>
#include <vector>

using namespace std;

// use DP(?)
// DP[i][j] = min distance of w1[:i+1] and w2[:j+1]
// we can use: insert/delete/replace
// if w1[i] == w2[j] => same as dp[i-1][j-1]
// else (e.g. hor, hoa) => min from:
//          replace(dp[i-1][j-1]+1)
//          insert(dp[i][j-1]+1, dp[i-1][j]+1)

// e.g. w1 = horse, w2 = ros
//    0  1  2  3  4  5    w1
// 0  1
// 1
// 2
// 3
// w2

class Solution {
public:
  int minDistance(string word1, string word2) {
    if (word1.size() == 0 && word2.size() == 0) return 0;
    if (word1.size() == 0) return word2.size();
    if (word2.size() == 0) return word1.size();

    int A = word1.size(), B = word2.size();
    vector<vector<int>> dp(A+1, vector<int>(B+1, 0));
    for (int i = 0; i <= A; ++i) {
      for (int j = 0; j <= B; ++j) {
        if (i==0 && j==0) {
          dp[i][j] = 0;               // edge case
        } else if (i == 0) {
          dp[i][j] = dp[i][j-1]+1;    // insert/delete
        } else if (j == 0) {
          dp[i][j] = dp[i-1][j]+1;    // insert/delete
        } else if (word1[i-1] == word2[j-1]) {
          dp[i][j] = dp[i-1][j-1];    // as same as (w1[i-2], w2[j-2])
        } else {
          dp[i][j] = min({dp[i-1][j-1], dp[i-1][j], dp[i][j-1]})+1; // min from replace, insert
        }
      }
    }
    return dp[A][B];
  }
  
};

// clang++ -std=c++17 c++/72.cpp -o 72 && ./72 -v
int main() {
  cout << Solution().minDistance("horse", "ros") << endl;
  cout << Solution().minDistance("intention", "execution") << endl;
  cout << Solution().minDistance("sea", "eat") << endl;
}